### PR TITLE
[BEAM-4679] Re-adding Flatten roots in ULR.

### PR DESCRIPTION
**Pull request by youngoli**
This was causing issues while testing pipelines with the ULR, so this
is a quick fix for that.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by youngoli** [comment]
R: @angoenka 
This fixes a few of the failing tests that popped up in the ValidatesRunner tests for the ULR.

**Comment by youngoli** [comment]
Run Java PreCommit

**Comment by angoenka** [comment]
Is this only needed for the tests or is it required for the correctness of the ReferenceRunner?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/EmptyInputProvider.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+   *
+   * <p>Returns an empty collection.
+   */
+  @Override
+  public Collection<CommittedBundle<Void>> getInitialInputs(
+      PTransformNode transform,
+      int targetParallelism) {
+    return Collections.emptyList();
+  }
+}
```
Comment (angoenka) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/EmptyInputProvider.java]: Please add an empty line

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/RootProviderRegistry.java [hunk 2]
```diff
@@ -33,6 +34,16 @@
...
+        ImmutableMap.<String, RootInputProvider<?>>builder()
+            .put(
+                IMPULSE_TRANSFORM_URN,
+                new ImpulseEvaluatorFactory.ImpulseRootProvider(bundleFactory))
+            .put(FLATTEN_TRANSFORM_URN, new EmptyInputProvider())
+        .build());
+  }
+
   /** Returns a {@link RootProviderRegistry} that only supports the {@link Impulse} primitive. */
   public static RootProviderRegistry impulseRegistry(BundleFactory bundleFactory) {
```
Comment (angoenka) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/RootProviderRegistry.java]: Shall we remove impulse registry as it is not used anymore?

**Comment by angoenka** [comment]
Please rerun `./gradlew  spotlessApply`

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/EmptyInputProvider.java (hunk 1 above)
Comment (youngoli) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/EmptyInputProvider.java]: Done. Spotless added it.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/RootProviderRegistry.java (hunk 2 above)
Comment (youngoli) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/RootProviderRegistry.java]: Agreed. Done.

**Comment by youngoli** [comment]
> Do we also need to add other primitive transforms here?
Reference: https://github.com/apache/beam/blob/master/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java#L133

I don't think that's necessary. This class only handles providers for roots of the pipeline. Theoretically only the ImpulseProvider should be necessary, but Java pipelines allow Flattens as roots, so to get the ValidatesRunner tests around that working this needed to be added.

Basically, I'm trying to avoid adding anything unless the ValidatesRunner tests fail otherwise.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/EmptyInputProvider.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+   * {@inheritDoc}.
+   *
+   * <p>Returns an empty collection.
+   */
+  @Override
+  public Collection<CommittedBundle<Void>> getInitialInputs(
+      PTransformNode transform, int targetParallelism) {
+    return Collections.emptyList();
+  }
+}
```
Comment (angoenka) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/EmptyInputProvider.java]: nit: new line. Spotless doesn't add it

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/RootProviderRegistry.java [hunk 4]
```diff
@@ -18,28 +18,28 @@
...
 import org.apache.beam.sdk.transforms.PTransform;
 
 /**
  * A {@link RootInputProvider} that delegates to primitive {@link RootInputProvider} implementations
  * based on the type of {@link PTransform} of the application.
  */
 class RootProviderRegistry {
-  /** Returns a {@link RootProviderRegistry} that only supports the {@link Impulse} primitive. */
-  public static RootProviderRegistry impulseRegistry(BundleFactory bundleFactory) {
+  static RootProviderRegistry javaPortableRegistry(BundleFactory bundleFactory) {
```
Comment (angoenka) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/RootProviderRegistry.java]: nit: java doc

Comment (youngoli) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/RootProviderRegistry.java]: Done, added the comment.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/EmptyInputProvider.java (hunk 3 above)
Comment (youngoli) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/EmptyInputProvider.java]: This is showing up without the line at the end of file on Github for me, but seems to actually have it in the repo. I think Github might just not be updating.

**Comment by youngoli** [comment]
Fixed up the nits and squashed the commits in preparation for merging.

**Comment by angoenka** [comment]
Thanks @youngoli 
We can merge this after the test cases pass.
cc: @tweise @lukecwik

**Comment by angoenka** [comment]
Failing test cases
org.apache.beam.sdk.io.elasticsearch.ElasticsearchIOTest.testWritePartialUpdate
org.apache.beam.sdk.io.elasticsearch.ElasticsearchIOTest.testWriteWithIdFn
Are unrelated to this PR.

**Comment by youngoli** [comment]
Rebasing to the most recent commit in master.

**Comment by angoenka** [comment]
Ping for merge
